### PR TITLE
Android: Rewrite image loading with Kotlin and Coil

### PR DESCRIPTION
**Pull request by t895**
When working on implementing the GameTDB fallbacks for PAL covers through glide, our code was getting bloated with boilerplate. Additionally, since we're starting to migrate to Kotlin, I thought this would be a good time to get all of the image loading code rewritten. Glide starts to feel outdated next to other Kotlin-native code so Coil was a natural choice. It is written for Kotlin first and we save a big volume of code taking this route. So now all image caching and fallbacks are handled gracefully by Coil. I've also taken the time to rewrite all of the files that work with image loading in Kotlin so we could take full advantage of coroutines and completely remove the messy set of Executors and Loopers from before. Everything responds nicely to the activity's lifecycle and we maintain the same performance.

Note - This also fixes behavior with the grid options where previously if you disabled downloading covers, the covers would just stay there. Now they are disabled outright and unless you have a custom cover, the no_cover image is displayed.

### CODE REVIEW

**Comment by JosJuice** [comment]
Would it be possible to first have a commit that translates to Kotlin and then a commit that actually makes the changes to how we load images? The PR would be a lot easier to review if so. Or would that be difficult to split up retroactively?

**Comment by t895** [comment]
> Would it be possible to first have a commit that translates to Kotlin and then a commit that actually makes the changes to how we load images? The PR would be a lot easier to review if so. Or would the changes be difficult to split retroactively?

It would bedifficult retroactively. Using Coil dependson using Kotlin in all of the classes that use it. So if I were to go back I'd have to rewrite it without coroutines and then with them.

At least to explain things better, instead of having GlideUtils that manages the threads and call each load like this 
`GlideUtils.loadGameCover(holder, holder.binding.imageGameScreen, gameFile, mActivity);`

We now bind the image loading to the lifecycle of the activity that spawns it. This means that when the activity is destroyed, all loading within this block is cancelled safely. The first portion of the block starts a load on the system-provided IO thread to find the custom cover and once that work is finished, we pass the result into the Coil load that works very similarly to how the Glide one would go.
`mActivity.lifecycleScope.launchWhenStarted {`
`            withContext(Dispatchers.IO) {`
`                val customCoverUri = CoilUtils.findCustomCover(gameFile)`
`                withContext(Dispatchers.Main) {`
`                    CoilUtils.loadGameCover(`
`                        holder,`
`                        holder.binding.imageGameScreen,`
`                        gameFile,`
`                        customCoverUri`
`                    )`
`                }`
`            }`
`        }`

**Comment by JosJuice** [comment]
> It would be difficult retroactively. Using Coil depends on using Kotlin in all of the classes that use it. So if I were to go back I'd have to rewrite it without coroutines and then with them.

I'm not asking for first a commit that makes us use Coil and then a commit that makes us use Kotlin, I'm asking for first a commit that makes us use Kotlin (while keeping Glide) and then a commit that makes us use Coil.

**Comment by t895** [comment]
> > It would be difficult retroactively. Using Coil depends on using Kotlin in all of the classes that use it. So if I were to go back I'd have to rewrite it without coroutines and then with them.
> 
> I'm not asking for first a commit that makes us use Coil and then a commit that makes us use Kotlin, I'm asking for first a commit that makes us use Kotlin (while keeping Glide) and then a commit that makes us use Coil.

Let me see what I can do

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/CoverHelper.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            4 -> "KO"
+            2 -> when (game.country) {
+                3 -> "AU"
+                4 -> "FR"
+                5 -> "DE"
+                6 -> "IT"
+                8 -> "NL"
+                9 -> "RU"
+                10 -> "ES"
+                0 -> "EN"
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/CoverHelper.kt:29]: Could we keep the comments here? Or alternatively, create named constants for the integers.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                binding.textGameTitle.text = gameFile.title
+                binding.textGameTitle.visibility = View.VISIBLE
+                binding.textGameTitleInner.visibility = View.GONE
+                binding.textGameCaption.visibility = View.VISIBLE
+            } else {
+                binding.textGameTitleInner.text = gameFile.title
+                binding.textGameTitle.visibility = View.GONE
+                binding.textGameCaption.visibility = View.GONE
+            }
+        }
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt:82]: This wasn't here before. If you want to move this, please do it outside of the first commit.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/GameDetailsDialog.kt [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            }
+
+            loadGameBanner(tvBinding.banner, gameFile)
+
+            builder.setView(tvBinding.root)
+        }
+        return builder.create()
+    }
+
+    fun loadGameBanner(imageView: ImageView, gameFile: GameFile) {
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/GameDetailsDialog.kt]: Same for this function.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameRowPresenter.kt [hunk 4]
```diff
@@ -7,20 +7,29 @@ import android.view.ViewGroup
...
 /**
  * The Leanback library / docs call this a Presenter, but it works very
  * similarly to a RecyclerView.Adapter.
  */
-class GameRowPresenter : Presenter() {
+class GameRowPresenter(activity: FragmentActivity) : Presenter() {
+    private val mActivity: FragmentActivity
+
+    init {
+        mActivity = activity
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameRowPresenter.kt:30]: I'm not all that familiar with Kotlin yet, but is this necessary? Can `activity` not be accessed outside of the `init` block?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/CoilUtils.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        } else if (BooleanSetting.MAIN_USE_GAME_COVERS.booleanGlobal) {
+            imageView.load(CoverHelper.buildGameTDBUrl(gameFile, CoverHelper.getRegion(gameFile))) {
+                error(R.drawable.no_banner)
+                target(
+                    onSuccess = { success ->
+                        disableInnerTitle(gameViewHolder)
+                        imageTarget.drawable = success
+                    },
+                    onError = { error ->
+                        if (gameFile.region == 2) {
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/CoilUtils.kt:49]: Please add a named constant in GameFile.java for this 2. The code is hard to follow otherwise.

**Review by JosJuice — COMMENTED** [review]
How is the caching handled now? I couldn't see any explicit mention of it.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameRowPresenter.kt (hunk 4 above)
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameRowPresenter.kt:30]: I looked this up in more detail. Seems like you can't use them just anywhere, but you should be able to use it when initializing `mActivity`, like this: `private val mActivity = activity`

Or, you could change the `activity` declaration to `val activity` to make it accessible throughout the class. (Though, should it be `private val activity` then...?)

Comment (t895) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameRowPresenter.kt:30]: I'll just go with `private val mActivity: FragmentActivity = activity` where applicable.

**Comment by t895** [comment]
> How is the caching handled now? I couldn't see any explicit mention of it.

It's handled internally by Coil. They maintain a memory and disk cache. It's customizable if we need later down the line too.

**Comment by JosJuice** [comment]
> It's handled internally by Coil. They maintain a memory and disk cache. It's customizable if we need later down the line too.

Does this mean that [the refetching problem](https://github.com/dolphin-emu/dolphin/pull/10626#discussion_r942867598) is back?

**Comment by t895** [comment]
> > It's handled internally by Coil. They maintain a memory and disk cache. It's customizable if we need later down the line too.
> 
> Does this mean that [the refetching problem](https://github.com/dolphin-emu/dolphin/pull/10626#discussion_r942867598) is back?

It is, but we talked about how preventing this issue isn't really reasonable to implement in IRC. Unless I misunderstood, I thought it was more reasonable that we deal with extra network requests.

Right now, if someone has a game that doesn't have the right cover for their language in PAL, it will default to the generic region and download that cover. But at least on subsequent tries, it will have one unnecessary network request and then load the cached fallback image.

**Comment by JosJuice** [comment]
> It is, but we talked about how preventing this issue isn't really reasonable to implement in IRC. Unless I misunderstood, I thought it was more reasonable that we deal with extra network requests.

I don't recall this.

> Right now, if someone has a game that doesn't have the right cover for their language in PAL, it will default to the generic region and download that cover. But at least on subsequent tries, it will have one unnecessary network request and then load the cached fallback image.

Yes, and that means that every time you open Dolphin from then on, Dolphin will do as many unnecessary network requests as you have affected games. I don't think that's reasonable.

**Comment by JosJuice** [comment]
For reference for others reading this: We agreed in PMs that dropping support for fallbacks would be fine for now.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import androidx.lifecycle.lifecycleScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.dolphinemu.dolphinemu.databinding.CardGameBinding
+import org.dolphinemu.dolphinemu.dialogs.GamePropertiesDialog
+import org.dolphinemu.dolphinemu.features.settings.model.BooleanSetting
+import org.dolphinemu.dolphinemu.utils.CoilUtils
+import java.util.ArrayList
+
+class GameAdapter(val mActivity: FragmentActivity) : RecyclerView.Adapter<GameViewHolder>(),
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt:27]: We should probably use `private` for attributes that were previously `private`.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt [hunk 7]
```diff
@@ -65,7 +58,32 @@ class GameAdapter(activity: Activity) : RecyclerView.Adapter<GameViewHolder>(),
...
+        holder.apply {
+            if (BooleanSetting.MAIN_SHOW_GAME_TITLES.booleanGlobal) {
+                binding.textGameTitle.text = gameFile.title
+                binding.textGameTitle.visibility = View.VISIBLE
+                binding.textGameTitleInner.visibility = View.GONE
+                binding.textGameCaption.visibility = View.VISIBLE
+            } else {
+                binding.textGameTitleInner.text = gameFile.title
+                binding.textGameTitle.visibility = View.GONE
+                binding.textGameCaption.visibility = View.GONE
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt:70]: You should probably explicitly set `binding.textGameTitleInner.visibility` to `View.VISIBLE` here, in case we're recycling a view that had it set to `View.GONE`.

Comment (t895) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/adapters/GameAdapter.kt:70]: Good catch
